tsl/src/remote/dist_copy.c

```diff
@@ -1248,355 +1225,369 @@ calculate_hyperspace_point_from_binary(Datum *values, bool *nulls, const Hypersp
-			 * chunk->data_nodes will point to the same list and no subsequent metadata updates will
-			 * occur.
-			 */
-			if (ht->fd.replication_factor > list_length(chunk_data_nodes))
-				ts_cm_functions->dist_update_stale_chunk_metadata(chunk, chunk_data_nodes);
-		}
+			rc = WaitEventSetWait(wset, -1, occurred, num_blocked, WAIT_EVENT_COPY_FILE_WRITE);
 
-		/*
-		 * For remote copy, we don't use chunk insert states on the AN.
-		 * So we need to explicitly set the chunk as partial when copies
-		 * are directed to previously compressed chunks.
-		 */
-		if (ts_chunk_is_compressed(chunk) && (!ts_chunk_is_partial(chunk)))
-			ts_chunk_set_partial(chunk);
-
-		/*
-		 * Schedule the row for sending to the data nodes containing the chunk.
-		 */
-		ListCell *lc;
-		foreach (lc, chunk->data_nodes)
-		{
-			ChunkDataNode *chunk_data_node = lfirst(lc);
-			/* Find the existing insert state for this data node. */
-			DataNodeRows *data_node_rows = NULL;
-			ListCell *lc2;
-			foreach (lc2, data_nodes)
+			for (i = 0; i < rc; i++)
 			{
-				data_node_rows = lfirst(lc2);
-				if (chunk_data_node->foreign_server_oid == data_node_rows->server_oid)
+				if (occurred[i].events & WL_LATCH_SET)
 				{
-					break;
+					ResetLatch(MyLatch);
+					CHECK_FOR_INTERRUPTS();
 				}
-			}
-
-			if (lc2 == NULL)
-			{
-				/* No insert state for this data node yet. Create it. */
-				data_node_rows = palloc(sizeof(DataNodeRows));
-				data_node_rows->connection = NULL;
-				data_node_rows->server_oid = chunk_data_node->foreign_server_oid;
-				data_node_rows->rows_total = 0;
-				/*
-				 * Not every tuple in a batch might be sent to every data node,
-				 * but we allocate the maximum possible size to avoid resizing.
-				 */
-				data_node_rows->row_indices = palloc(sizeof(int) * context->batch_row_count);
-				data_nodes = lappend(data_nodes, data_node_rows);
-			}
-
-			Assert(data_node_rows->server_oid == chunk_data_node->foreign_server_oid);
-
-			data_node_rows->row_indices[data_node_rows->rows_total] = row_in_batch;
-			data_node_rows->rows_total++;
-		}
-	}
+				else if (occurred[i].events & WL_SOCKET_WRITEABLE)
+				{
+					DataNodeConnection *dnc = occurred[i].user_data;
+					PGconn *pg_conn = remote_connection_get_pg_conn(dnc->connection);
 
-	/*
-	 * Flush the previous batch to avoid growing the outgoing buffers
-	 * indefinitely if some data node is not keeping up. It would be more
-	 * efficient to check for buffer growth and only flush then, but libpq
-	 * doesn't provide a way to know the outgoing buffer size. It also doesn't
-	 * provide any way to control the outgoing buffer size.
-	 * Don't do it if we have ended the COPY above to create new chunks.
-	 * The number 11 is an arbitrary prime, growing the output buffer to at
-	 * most 11ki rows sounds reasonable.
-	 */
-	if (!did_end_copy && context->batch_ordinal % 11 == 0)
-	{
-		flush_active_connections(&context->connection_state);
-	}
```

Comment:
Do we still do the periodic flush somewhere? We need some way to control the outgoing queue size. I expect we're going to be bottlenecked at the data nodes in the normal case, so we need to avoid growing the output buffers on access node indefinitely.

Reply:
So, I removed this because I don't think it is necessary.

The reason I don't think it is necessary is because even if we'd use standard libpq functions, it actually flushes automatically once a msg end marker is written and the buffer has reached 8k. 

Further, now that I changed this to use the internal API, we have better control on a per socket basis. So, every socket is flushed once we write the CopyData end marker, which happens after N rows are buffered. So, basically, we accumulate up to N rows of data on a socket, then write msg end and flush. If the flush fails, the code will do a WaitEventSetWait on all sockets that could not be flushed.

Previously, we flushed all sockets after having sent N rows across all sockets, which meant that some sockets might not require a flush while some does.

I kept the `flush_active_connections()` at the end of the transfer, but it only does something if any of the sockets failed the their last flush for some reason. Then it will wait until the flush succeeds.

Reply:
> So, every socket is flushed once we write the CopyData end marker, which happens after N rows are buffered. So, basically, we accumulate up to N rows of data on a socket, then write msg end and flush. If the flush fails, the code will do a WaitEventSetWait on all sockets that could not be flushed.

Does this guarantee that we're flushing at the same rate we're receiving the new data, and the buffer size doesn't grow? Before it just flushed to zero outgoing buffers, now that you imported the internal header, probably we can just examine the buffer size.

Reply:
After reading the code more, looks like `send_row_to_data_nodes` flushes to zero output buffer, right?

Reply:
Thanks, yes, it's more clear.